### PR TITLE
rework of DE translation of editor-client/locales/de/jsonata.json

### PR DESCRIPTION
**Pull request by heikokue**
- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## Proposed changes
complete rework of DE translation of editor-client/locales/de/jsonata.json
see https://discourse.nodered.org/t/german-translation/5366
and https://node-red.slack.com/archives/CK09P5RHR

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [ ] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality
- [x] **I checked them on a RaspberryPi life system**

### CODE REVIEW

**Comment by knolleary** [comment]
Looks like this is based on the `dev` branch, so the pr needs to be retargeted to that branch, not `master`.

You can edit the target branch by clicking the edit button next to the PR title.

**Comment by heikokue** [comment]
> Looks like this is based on the `dev` branch, so the pr needs to be retargeted to that branch, not `master`.
> You can edit the target branch by clicking the edit button next to the PR title.

done

**Comment by JoergWende** [comment]
I'm going to look at the translation if you want ...
J.

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json [hunk 1]
```diff
@@ -1,222 +1,274 @@
...
+    "args": "number, picture",
+    "desc": "Wandelt `number` in eine Zeichenfolge um und formatiert sie in einer Ganzzahl-Darstellung, spezifiziert durch den `picture`-String-Parameter. Der `picture`-String-Parameter definiert, wie die Zahl `number` formatiert werden soll und hat den selben Syntax wie `fn:format-integer` der XPath F&O 3.1 Spezifikation."
+  },
+  "$parseInteger": {
+    "args": "string, picture",
+    "desc": "Wandelt den Inhalt von `string` in eine Ganzzahl `integer` (als JSON Zahl), spezifiziert durch den `picture`-String-Parameter. Der `picture`-String-Parameter hat das selbe Format wie `$formatInteger`."
+  },
+  "$error": {
+    "args": "[str]",
+    "desc": "Meldet ein Fehler mit einer Meldung. Der optionale String `str` ersetzt die Standardmeldung `$error() function evaluated`."
```
Comment (HLeithner) [inline packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:236]:
```suggestion
    "desc": "Erzeugt eine Fehlermeldung. Der optionale String `str` ersetzt die Standardmeldung `$error() function evaluated`."
```

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json [hunk 2]
```diff
@@ -1,222 +1,274 @@
...
+  },
+  "$assert": {
+    "args": "arg, str",
+    "desc": "Wenn `arg` gleich `true` ist, liefert die Function `undefined` zurück. Wenn `arg` gleich `false` ist, wird ein Ausnahmefehler gemeldet mit dem String_Parameter `str` als Meldetext."
+  },
+  "$single": {
+    "args": "array, function",
+    "desc": "Gibt den einzigen Wert im Wert `array` zurück, der die Bedingung `function` erfüllt (d.h. die Funktion `function` gibt den boole'schen Wert `true` zurück, wenn der Wert übergeben werden soll). Sie meldet einen Ausnahmefehler, wenn die Anzahl der übergebenen Werte nicht genau eins ist.\n\nDie Funktion `function` sollte in der folgenden Art angegeben werden: `function(value [, index [, array]])` wobei `value` für jeden Eintrag des Arrays gilt, `index` ist die Position dieses Eintrags und das gesamte Array `array` wird als dritter Parameter übergeben."
```
Comment (HLeithner) [inline packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:244]:
```suggestion
    "desc": "Gibt ein einziges Element aus `array` zurück, welches die Bedingung `function` erfüllt (d.h. die Funktion `function` gibt den boole'schen Wert `true` zurück, wenn der Wert übergeben werden soll). Sie meldet einen Ausnahmefehler, wenn die Anzahl der übergebenen Werte nicht genau eins ist.\n\nDie Funktion `function` sollte in der folgenden Art angegeben werden: `function(value [, index [, array]])` wobei `value` für jedes Element des Arrays gilt, `index` ist die Position dieses Elements und das gesamte Array `array` wird als dritter Parameter übergeben."
```

**Review by HLeithner — COMMENTED** [review]
I added some improvements maybe you find them useful.

**Comment by JoergWende** [comment]
Hi Harald,
glad to have you here - please lets coordinate the PR's on slack
Cheers
Joerg

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json [hunk 3]
```diff
@@ -1,222 +1,274 @@
...
   "$length" : {
     "args" : "str",
-    "desc" : "Gibt die Anzahl der Zeichen in der Zeichenfolge `str` zurück. Es wird ein Fehler ausgelöst, wenn `str` keine Zeichenfolge ist."
+    "desc" : "Gibt die Zeichenanzahl von `str` zurück. Es wird ein Fehler ausgelöst, wenn `str` keine Zeichenfolge ist."
   },
   "$substring" : {
-    "args" : "str, start [, länge]",
-    "desc" : "Gibt eine Zeichenfolge zurück, die die Zeichen im ersten Parameter `str` beginnend bei Position `start` (Null-Offset) enthält. Wenn \"length\" angegeben ist, enthält die Unterzeichenfolge maximal \"Länge\" Zeichen. Wenn `start` negativ ist, gibt es die Anzahl der Zeichen am Ende von `str`  an."
+    "args" : "str, start [, length]",
+    "desc" : "Gibt eine Unterzeichenfolge zurück, die die Zeichen in `str` beginnend bei Position `start` (Null-Offset) enthält. Wenn `length` angegeben ist, enthält die rückgegebene Zeichenfolge maximal die in `length` angegebene Zeichenanzahl. Wenn `start` negativ ist, werden die Zeichen vom Ende aus gezählt von `str` zurückgegeben."
```
[the remark this answers is not in the document]
Comment (heikokue) [reply packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:12]: noch besser: Teilzeichenfolge

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json [hunk 4]
```diff
@@ -1,222 +1,274 @@
...
+    "args" : "array [, separator]",
+    "desc" : "Verkettet ein Array von Zeichenfolgen in eine einzige Zeichenfolge, wobei die einzelnen Zeichenfolgen durch den optionalen Trennzeichen-Parameter `separator` getrennt sind. Es ergibt einen Fehler, wenn das `array` ein Element enthält, das keine Zeichenfolge ist. Wenn `separator` nicht angegeben wird, wird davon ausgegangen, dass es sich um eine leere Zeichenfolge handelt, d.h. zwischen den einzelnen Zeichenfolgen wird kein Trennzeichen eingefügt. Es handelt sich um einen Fehler, wenn `separator` keine Zeichenfolge ist."
   },
   "$match" : {
-    "args" : "str, Muster [, Grenzwert]",
-    "desc" : "Wendet die Zeichenfolge `str`  an den regulären Ausdruck `Muster` an und gibt ein Array von Objekten zurück, wobei jedes Objekt Informationen zu jedem Vorkommen einer Übereinstimmung in `str` enthält."
+    "args" : "str, pattern [, limit]",
+    "desc" : "Wendet den regulären Ausdruck `pattern` bei der Zeichenfolge `str` an und gibt ein Array von Objekten zurück, die Informationen zu jedem Vorkommen von `pattern` in `str` enthält."
```
[the remark this answers is not in the document]
Comment (heikokue) [reply packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:48]: jo!

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json [hunk 5]
```diff
@@ -1,222 +1,274 @@
...
+    "args" : "array",
+    "desc" : "Gibt den Mittelwert eines `array` von Zahlen zurück. Es ergibt einen Fehler, wenn `array` ein Element enthält, das keine Zahl ist."
   },
   "$boolean" : {
     "args" : "arg",
-    "desc" : "Castet das Argument mit den folgenden Regeln in einen Booleschen Wert:\n\n -` Boolean ': nicht geändert\n -` string `: leer: `false`\n -` string `: nicht leer: `true`\n -` Zahl `: ` 0 `: ` falsch `\n -` Zahl `: Nicht-Null: `true`\n -` null `: `false`\n -` array `: leer: `false`\n -` array `: enthält ein Mitglied, das auf `true` setzt: `true`\n -` array `: alle Member werden in `false` umgesetzt: `false`\n -` object `: empty: `false`\n -` object `: non-empty: `true`\n -` Funktion `: ` falsch `"
+    "desc" : "Wandelt `arg` gemäß folgender Regeln in einen boole'schen Wert um:\n\n- `Boolean`: unverändert\n- `string`: leer → `false`, nicht leer → `true`\n- `Zahl`: `0` → `falsch`, Nicht-Null → `true`\n- `null` → `false`\n- `array`: leer → `false`, enthält mindestens ein Mitglied, das `true` ist → `true`, alle Mitglieder sind `false` → `false`\n- `object`: leer → `false`, nicht leer → `true`\n- `function`: `false`"
```
[the remark this answers is not in the document]
Comment (heikokue) [reply packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:120]: jo!

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json (hunk 1 above)
Comment (heikokue) [reply packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:236]: jo!

File: packages/node_modules/@node-red/editor-client/locales/de/jsonata.json (hunk 2 above)
Comment (heikokue) [reply packages/node_modules/@node-red/editor-client/locales/de/jsonata.json:244]: hier war noch einiges im Argen. Bitte nochmal genau prüfen!

**Comment by heikokue** [comment]
Hello Harald,
thanks for your review.
I fixed or reworked all of them.
Grüße,
Heiko

**Comment by HLeithner** [comment]
In my opinion in the translation are some structural problems, for example array elements have different "names" like "eintrag", "element", "wert", ... That's something that should be normalized.

Beside that it would be interesting who is the target group of this documentation, for me as developer it's ok to read it but it's not consequent addressing developer. Some descriptions are targeting (at least that's my feeling) end user without any programming knowledge. 

@JoergWende sorry I don't have slack access.

**Comment by JoergWende** [comment]
> In my opinion in the translation are some structural problems, for example array elements have different "names" like "eintrag", "element", "wert", ... That's something that should be normalized.
>
good point  
> Beside that it would be interesting who is the target group of this documentation, for me as developer it's ok to read it but it's not consequent addressing developer. Some descriptions are targeting (at least that's my feeling) end user without any programming knowledge.
> 
end users are the target group - that should be in our mind - so yes - you are correct
> @JoergWende sorry I don't have slack access.

Please check this link: https://app.slack.com/client/T03M2TANY/C03M2TAQ8 - there is a channel called #i18n-de. We used this channel in the past to discuss translation issues.

**Comment by heikokue** [comment]
Good morning!
@all: Thanks for your review!

> In my opinion in the translation are some structural problems, for example array elements have different "names" like "eintrag", "element", "wert", ... That's something that should be normalized.

I just pushed an optimized version using "Element" only in array descriptions.
Please review.
 
> Beside that it would be interesting who is the target group of this documentation, for me as developer it's ok to read it but it's not consequent addressing developer. Some descriptions are targeting (at least that's my feeling) end user without any programming knowledge.

Yes, I agree, the end user with a minimum technical experience (but without a degree) is the target reader of the docs.
My current primary goal is the translation en-US to de (only).
But if we want to improve the docs, whe should do this not only in the german version, but in all languages.

**Comment by dceejay** [comment]
Thanks for raising all these PR - I can see the discussion is getting lively and hopefully productive. 
Are you able to sign/click the CLA such that we are able to merge them as and when they come to a consensus ?

**Comment by heikokue** [comment]
merged into #2806

**Pull request closed without merging**